### PR TITLE
Feat/switch/language

### PR DESCRIPTION
**Pull request by Palashx**
New mechanism for adding, models to a project, and switching different models in a project was added. Also altered all the test, according to the new mechanism

### CODE REVIEW

File: botfront/cypress/integration/nlu_models/nlu_models.spec.js [hunk 1]
```diff
@@ -1,143 +1,75 @@
...
-            cy.get('[data-cy=offline-model]').click();
-            cy.get('[data-cy=confirm-popup]').contains('No').click();
-            cy.get('[data-cy=confirm-popup]').should('not.exist');
-            checkPublishingState(0, 1);
+    it('should NOT be able to delete the default model', function() {
+        cy.visit(`/project/${this.bf_project_id}/settings`);
+        cy.get('.project-default-language .ui > .search').click();
+        cy.get('.project-default-language input').type('English');
+        cy.get('.project-default-language').contains('English').click();
+        cy.get('.project-settings-menu-info').click();
```
Comment (znat) [inline botfront/cypress/integration/nlu_models/nlu_models.spec.js]: Should use [data-cy...] directives instead of classes if possible

File: botfront/cypress/integration/nlu_models/nlu_models.spec.js [hunk 2]
```diff
@@ -1,143 +1,75 @@
...
-            cy.visit(`/project/${this.bf_project_id}/nlu/models`);
-            cy.contains('French').click();
-            cy.get(':nth-child(1) > .extra > .basic > .primary').click();
-            cy.get('.nlu-menu-settings').click();
-            cy.contains('Import').click();
-            cy.fixture('nlu_import.json', 'utf8').then((content) => {
-                cy.get('.file-dropzone').upload(content, 'data.json');
-            });
-
-            cy.contains('Import Training Data').click();
```
Comment (znat) [inline botfront/cypress/integration/nlu_models/nlu_models.spec.js]: Why is this test removed?

File: botfront/cypress/integration/nlu_models/nlu_models.spec.js [hunk 3]
```diff
@@ -1,143 +1,75 @@
...
-            cy.contains('Delete').click();
-            cy.get('.nlu-menu-settings').click();
-            cy.get('.delete-model-button').should('be.disabled');
-            cy.get('.dowload-model-backup-button').click();
-            cy.get('.delete-model-button').should('be.enabled');
-            cy.get('.delete-model-button').click();
-            cy.get('.ui.page.modals').should('be.visible');
-            cy.get('.ui.page.modals .primary').click();
-            cy.url('should.be', `/project/${this.bf_project_id}/nlu/models`);
-            cy.get('.s-alert-success').should('be.visible');
```
Comment (znat) [inline botfront/cypress/integration/nlu_models/nlu_models.spec.js]: Why is model deletion not tested anymore?

File: botfront/cypress/integration/nlu_tagging/nlu_tagging.spec.js [hunk 4]
```diff
@@ -12,41 +12,49 @@
 
 const utterance = 'can you recover your asset in case of client default?';
```
Comment (znat) [inline botfront/cypress/integration/nlu_tagging/nlu_tagging.spec.js]: @MatthieuJnon we need to change the dataset.

File: botfront/cypress/integration/project_settings/project.settings.info.spec.js [hunk 5]
```diff
@@ -22,7 +22,7 @@ describe('Project Settings', function() {
             cy.visit(`/project/${this.bf_project_id}/settings`);
             cy.get('.project-name input').click();
             cy.get('.project-name input').type('33');
-            cy.get('.save-project-info-button').click();
+            cy.contains('Save Changes').click();
```
Comment (znat) [inline botfront/cypress/integration/project_settings/project.settings.info.spec.js]: prefer data-cy if possible. We should only use "contains" for dropdown because it's driven by data. Otherwise if we change the text it will break the test

File: botfront/cypress/support/index.js [hunk 6]
```diff
@@ -74,20 +74,44 @@ Cypress.Commands.add('createNLUModel', (projectId, name, language, description,
...
+                cy.visit(`/project/${id}/nlu/model/${modelId}`);
+            });
+            cy.get('.nlu-menu-settings').click();
+            cy.contains('Import').click();
+            cy.fixture('nlu_import.json', 'utf8').then((content) => {
+                cy.get('.file-dropzone').upload(content, 'data.json');
+            });
+        
+            cy.contains('Import Training Data').click();
+        });
```
Comment (znat) [inline botfront/cypress/support/index.js]: If I read well, you created a command to create a model from a JSON file, but remove the UI test for imports. We should still verify that importing data works. (see comment above)

File: botfront/imports/api/nlu_model/nlu_model.methods.js [hunk 7]
```diff
@@ -165,15 +172,34 @@ if (Meteor.isServer) {
...
+            if (modelLanguage.language !== projectDefaultLanguage.defaultLanguage) {
+                try {
+                    NLUModels.remove({ _id: modelId });
+                    Projects.update({ _id: projectId }, { $pull: { nlu_models: modelId } });
+                    return 'Model Deleted';
+                } catch (e) {
+                    throw e;
+                }
+            }
+            return 'Model Not Deleted';
```
Comment (znat) [inline botfront/imports/api/nlu_model/nlu_model.methods.js]: 1) Why does it return a String? Should return a Meteor.Error
2) This message is not helpful
You could `throw Meteor.Error('409', 'The default language cannot be deleted')`

File: botfront/imports/api/nlu_model/nlu_model.methods.js [hunk 8]
```diff
@@ -165,15 +172,34 @@ if (Meteor.isServer) {
...
             check(projectId, String);
             checkIfCan('nlu-admin', getProjectIdFromModelId(modelId));
+            // check the default language of project and the language of model
+            const modelLanguage = NLUModels.findOne({ _id: modelId });
+            const projectDefaultLanguage = Projects.findOne({ _id: projectId });
+            if (modelLanguage.language !== projectDefaultLanguage.defaultLanguage) {
+                try {
+                    NLUModels.remove({ _id: modelId });
+                    Projects.update({ _id: projectId }, { $pull: { nlu_models: modelId } });
+                    return 'Model Deleted';
```
Comment (znat) [inline botfront/imports/api/nlu_model/nlu_model.methods.js]: Don't return text, you can add a `return` statement at the beginning of l181

File: botfront/imports/api/nlu_model/nlu_model.methods.js [hunk 9]
```diff
@@ -165,15 +172,34 @@ if (Meteor.isServer) {
...
+            const { _id: modelId } = NLUModels.findOne({ _id: { $in: modelIds }, language });
             try {
                 NLUModels.remove({ _id: modelId });
                 Projects.update({ _id: projectId }, { $pull: { nlu_models: modelId } });
             } catch (e) {
                 throw e;
             }
-
-            return 'Model Deleted';
+            return `Model with language ${language} deleted.`;
```
Comment (znat) [inline botfront/imports/api/nlu_model/nlu_model.methods.js]: Save as above

File: botfront/imports/ui/components/nlu/import-export/DeleteModel.jsx [hunk 10]
```diff
@@ -44,9 +44,31 @@ export default class DeleteModel extends React.Component {
         this.setState({ backupDownloaded: true });
     };
 
+    renderCannotDeleteMessage = (cannotDelete) => {
+        if (!cannotDelete) {
+            return (
+                <Message
+                    header='Default language cannot be deleted'
+                    icon='warning'
+                    content={'You can\'t delete the default language, to delete this langauge change the default langauge of hte project.'}
```
Comment (znat) [inline botfront/imports/ui/components/nlu/import-export/DeleteModel.jsx]: Please fix the typo

Comment (Palashx) [reply botfront/imports/ui/components/nlu/import-export/DeleteModel.jsx]: Will fix it

File: botfront/imports/api/nlu_model/nlu_model.methods.js (hunk 9 above)
Comment (Palashx) [reply botfront/imports/api/nlu_model/nlu_model.methods.js]: Will fix it

File: botfront/imports/api/nlu_model/nlu_model.methods.js (hunk 7 above)
Comment (Palashx) [reply botfront/imports/api/nlu_model/nlu_model.methods.js]: Will fix it

File: botfront/cypress/support/index.js (hunk 6 above)
Comment (Palashx) [reply botfront/cypress/support/index.js]: Ok then I will add it again.

File: botfront/cypress/integration/project_settings/project.settings.info.spec.js (hunk 5 above)
Comment (Palashx) [reply botfront/cypress/integration/project_settings/project.settings.info.spec.js]: Will come up with data-cy

File: botfront/cypress/integration/nlu_models/nlu_models.spec.js (hunk 3 above)
Comment (Palashx) [reply botfront/cypress/integration/nlu_models/nlu_models.spec.js]: Will add this back

File: botfront/cypress/integration/nlu_models/nlu_models.spec.js (hunk 2 above)
Comment (Palashx) [reply botfront/cypress/integration/nlu_models/nlu_models.spec.js]: I removed it now because in several other tests a createNLUModelWithImport method is called, which does the exact same thing. So I thought it will be redundant to use have it again. I will add it again

File: botfront/cypress/integration/nlu_models/nlu_models.spec.js (hunk 1 above)
Comment (Palashx) [reply botfront/cypress/integration/nlu_models/nlu_models.spec.js]: Will try to use data-cy

File: botfront/cypress/integration/chat/chat.spec.js [hunk 11]
```diff
@@ -13,40 +21,18 @@ describe('chat side panel handling', function() {
...
-        cy.visit(`/project/${this.bf_project_id}/nlu/models`);
-        cy.get('[data-cy=open-chat]').click();
-        cy.get('[data-cy=no-language]');
-        cy.get('[data-cy=chat-language-option]').should('not.exist');
-        cy.get('[data-cy=close-chat]').click();
-
-        cy.get('[data-cy=offline-model]').click();
-        cy.get('[data-cy=confirm-popup]').contains('Yes').click();
-    });
-
```
[the remark this answers is not in the document]
Comment (Palashx) [reply botfront/cypress/integration/chat/chat.spec.js]: Yes the default language is always present

File: botfront/cypress/integration/bot_responses/bot_response.spec.js [hunk 12]
```diff
@@ -26,15 +23,6 @@ describe('Bot responses', function() {
...
         cy.fixture('bf_project_id.txt').as('bf_project_id');
-        cy.get('@bf_project_id').then((id) => {
-            cy.createNLUModel(id, modelName, modelLang, 'my description');
-        });
-    });
-
-    after(function() {
-        cy.login();
-        cy.deleteNLUModel(this.bf_project_id, modelName, modelLang);
-        cy.logout();
```
[the remark this answers is not in the document]
Comment (Palashx) [reply botfront/cypress/integration/bot_responses/bot_response.spec.js]: Will look in to it and make it atomic

File: botfront/cypress/integration/nlu_models/nlu_models.spec.js (hunk 3 above)
Comment (Palashx) [reply botfront/cypress/integration/nlu_models/nlu_models.spec.js]: It is already there in this test 'should be able to create a new Model by adding a new language, to the project and then delete it'

File: botfront/imports/ui/components/nlu/models/NLUModel.jsx [hunk 13]
```diff
@@ -165,35 +164,44 @@ class NLUModel extends React.Component {
         if (instance) {
             tabs.push({
                 menuItem: 'API',
-                render: () => (
-                    <React.Suspense fallback={<Loader active />}>
-                        <API model={model} instance={instance} />
-                    </React.Suspense>
-                ),
+                render: () => (<API model={model} instance={instance} />),
```
Comment (MatthieuJnon) [inline botfront/imports/ui/components/nlu/models/NLUModel.jsx]: Why did you remove the React Suspense here ?

File: botfront/imports/ui/components/nlu/models/NLUModel.jsx [hunk 14]
```diff
@@ -165,35 +164,44 @@ class NLUModel extends React.Component {
...
-            { menuItem: 'General', render: () => <NLUParams model={model} instances={instances} /> },
             { menuItem: 'Pipeline', render: () => <NLUPipeline model={model} onSave={this.onUpdateModel} /> },
             { menuItem: 'Import', render: () => <DataImport model={model} /> },
             { menuItem: 'Export', render: () => <DataExport intents={intents} model={model} /> },
-            { menuItem: 'Delete', render: () => <DeleteModel model={model} onDeleteModel={this.onDeleteModel} /> },
+            { menuItem: 'Delete', render: () => <DeleteModel model={model} onDeleteModel={this.onDeleteModel} cannotDelete={cannotDelete} /> },
         ];
     };
 
+    onChange = (e, data) => {
```
Comment (MatthieuJnon) [inline botfront/imports/ui/components/nlu/models/NLUModel.jsx]: don't just call it onChange, we don't know what's changing.
Methods that update controlled components should be called handle<something>
Here you could call the method handleLanguageChange() for instance

File: botfront/imports/ui/components/nlu/models/NLUModel.jsx [hunk 15]
```diff
@@ -263,7 +271,7 @@ class NLUModel extends React.Component {
         return (
             <div id='nlu-model'>
                 <Menu pointing secondary>
-                    <Menu.Item>{this.getHeader()}</Menu.Item>
+                    <Menu.Item style={{ paddingBottom: '0.3rem' }} header>{this.getHeader()}</Menu.Item>
```
Comment (MatthieuJnon) [inline botfront/imports/ui/components/nlu/models/NLUModel.jsx]: Don't add style in style node property, it's better to write it in a .less file because Meteor will automatically optimize code in those files.
You probably don't have to add a class, just write a css query specific enough

File: botfront/imports/ui/components/nlu/models/NLUModel.jsx [hunk 16]
```diff
@@ -337,27 +345,50 @@ class NLUModel extends React.Component {
...
     intents: [],
+    ready: false,
+    nluModelLanguages: [],
+    models: [],
+    projectDefaultLanguage: '',
+    projectId: '',
+    model: [],
 };
 
 const NLUDataLoaderContainer = withTracker((props) => {
```
Comment (MatthieuJnon) [inline botfront/imports/ui/components/nlu/models/NLUModel.jsx]: Maybe try to split up this method into smaller function
(You can put the functions directly into the method so as to not pollute the upper namespace)

File: botfront/cypress/integration/nlu_models/nlu_models.spec.js [hunk 17]
```diff
@@ -1,143 +1,92 @@
...
+    it('should be able to call the nlu.remove Meteor Call for the any model (not default)', function() {
```
[the remark this answers is not in the document]
Comment (Palashx) [reply botfront/cypress/integration/nlu_models/nlu_models.spec.js]: Will remove the meteor call for deleting.

File: botfront/cypress/integration/nlu_models/nlu_models.spec.js [hunk 18]
```diff
@@ -1,143 +1,92 @@
...
-            cy.get('[data-cy=duplicate-button]').click();
-            cy.get('[data-cy=confirm-popup]').should('not.exist');
-            cy.get('[data-cy=duplicate-button]').click();
-            cy.get('[data-cy=confirm-popup]').contains('No').click();
-            cy.get('[data-cy=nlu-model-card]').should('have.lengthOf', 1);
-            cy.get('[data-cy=duplicate-button]').click();
-            cy.get('[data-cy=confirm-popup]').contains('Yes').click();
-            cy.get('[data-cy=nlu-model-card]').should('have.lengthOf', 2);
-            deleteSecondModel();
+    it('Should import a json file with training data', function() {
```
[the remark this answers is not in the document]
Comment (Palashx) [reply botfront/cypress/integration/nlu_models/nlu_models.spec.js]: Will remove this test, and it also implicitly tested in index.js.

File: botfront/imports/api/instances/instances.methods.js [hunk 19]
```diff
@@ -20,7 +20,8 @@ export const createInstance = async (project) => {
             return;
         }
         if (instance) {
-            Instances.insert(instance);
+            // eslint-disable-next-line consistent-return
+            return await Instances.insert(instance);
```
[the remark this answers is not in the document]
Comment (Palashx) [reply botfront/imports/api/instances/instances.methods.js]: It returns the id of the instance created, which is used by the project.insert method to associate the instance id with the project.

File: botfront/imports/api/nlu_model/nlu_model.methods.js [hunk 20]
```diff
@@ -140,7 +140,14 @@ if (Meteor.isServer) {
         'nlu.insert'(item, projectId) {
             check(item, Object);
             check(projectId, String);
-
+            // Check if the model with the langauge already exists in project
+            // eslint-disable-next-line no-param-reassign
+            item.published = true; // a model should be published as soon as it is created
+            const { nlu_models } = Projects.findOne({ _id: projectId }, { fields: { nlu_models: 1 } });
```
[the remark this answers is not in the document]
Comment (Palashx) [reply botfront/imports/api/nlu_model/nlu_model.methods.js]: Solved

File: botfront/imports/api/nlu_model/nlu_model.methods.js [hunk 21]
```diff
@@ -140,7 +140,14 @@ if (Meteor.isServer) {
...
+            const nluModelLanguages = getNluModelLanguages(nlu_models, true);
+            if (nluModelLanguages.map(lang => (lang.value)).includes(item.language)) {
```
[the remark this answers is not in the document]
Comment (Palashx) [reply botfront/imports/api/nlu_model/nlu_model.methods.js]: Done

File: botfront/imports/api/nlu_model/nlu_model.methods.js [hunk 22]
```diff
@@ -359,10 +383,8 @@ if (Meteor.isServer) {
                     fr: JSON.parse(Assets.getText('nlu/nlu-chitchat-fr.json')),
                     en: JSON.parse(Assets.getText('nlu/nlu-chitchat-en.json')),
                 };
-                const projectId = await Meteor.callWithPromise('project.insert', { name: 'Chitchat', _id: `chitchat-${shortid.generate()}`, namespace: 'chitchat' });
+                const projectId = await Meteor.callWithPromise('project.insert', { name: 'Chitchat', _id: `chitchat-${shortid.generate()}`, namespace: 'chitchat', defaultLanguage: 'en' });
```
[the remark this answers is not in the document]
Comment (Palashx) [reply botfront/imports/api/nlu_model/nlu_model.methods.js]: Solved

File: botfront/imports/ui/components/nlu/models/NLUModel.jsx (hunk 13 above)
Comment (Palashx) [reply botfront/imports/ui/components/nlu/models/NLUModel.jsx]: App was crashing because of some React.lazy not working properly. Nathan told to leave it for now and we will fix it later. So I removed it.

File: botfront/imports/ui/components/nlu/models/NLUModel.jsx (hunk 14 above)
Comment (Palashx) [reply botfront/imports/ui/components/nlu/models/NLUModel.jsx]: Done, I forgot about the convention. Changed it to handleLanguageChange

File: botfront/imports/ui/components/nlu/models/NLUModel.jsx (hunk 15 above)
Comment (Palashx) [reply botfront/imports/ui/components/nlu/models/NLUModel.jsx]: Done

File: botfront/imports/ui/components/nlu/models/NLUModel.jsx (hunk 16 above)
Comment (Palashx) [reply botfront/imports/ui/components/nlu/models/NLUModel.jsx]: Tried to split but most of it is very inter connected. But still added some splitting of function
